---
title: "Practical I: epidemiological surveillance and modelling (non-spatial)"
date: "25/26th January 2024"
output:
  pdf_document: default
  html_document:
    df_print: paged
keep_tex: yes
fig_caption: yes
geometry: margin=1in
fontsize: 11pt
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message=FALSE, warning=FALSE, fig.path='Figs/', fig.width=10, fig.height=4, root.dir = '')
```



  In this first practical the student is expected to learn to download, visualise, analyse, and interpret epidemiological case count data. Further, the student is expected to reflect on the issues these data present and provide a perspective on how to improve disease surveillance and analytics in order to better estimate basic epidemiological parameters.

The practical will start with a short 15-20 minute presentation by the course convenor and demonstrators. The next 2h will be spent answering predefined questions before a short wrap up and reflection at the end.

There will be a 15 minute break after ca. 1h ½.

```{r Load required packages,echo=FALSE}
# Required packages for practicle
if (!require("pacman")) install.packages("pacman")
pkgs =
  c("knitr", "rmarkdown", "pander", "EpiEstim",
    "ggplot2", "incidence", "dplyr") # End
pacman::p_load(pkgs, character.only = T)

# remove unwanted objects (package list)
rm(pkgs)

```


## Practical I Questions

\newpage

#### Q1: What are common disease surveillance data and what are their pros and cons? Name 4 at most. (ca. 15 minutes)

<p>&nbsp;</p>

Model answer:

```{r Create table,echo=FALSE}
library(knitr)
library(rmarkdown)
library(pander)

mytable = data.frame(
    "Data Type" = c("Syndromic surveillance", "Genomic surveillance","Serological surveillance", "Wastewater surveillance"),
    Pros = c("* Early detection\ \n* Relatively cheap", "* Detailed pathogen information\ \n* Allows for downstream phylogenetic analysis and tracking of evolution (e.g. Drug resistance and Mutations)\ \n*  Identify new/emerging pathogens", "* Overall burden of disease in population\ \n* Past exposure to pathogen(s)", "* Early detection\ \n* Can measure overall disease burden\ \n* Cost-effective + non-invasive"),
    Cons     = c("* Higher chances for misdiagnosis\ \n* Little information on pathogen evolution","* Relatively expensive \ \n* Requires special expertise\ \n* Tends to be slower at detecting new outbreaks", "* May not reflect current infections\ \n* Susceptible to cross-reactivity\ \n* Expensive (REACT, ONS Surveys)", "* Difficult to identify individuals infected"))

pander::pander(mytable, keep.line.breaks = TRUE, style = 'grid', justify = 'left')
```


#### Q2: What are some of the challenges with public health surveillance at the beginning of disease outbreaks? (ca. 5 minutes)
<p>&nbsp;</p>

Model answer:

- Speed of collection, cleaning and sharing of data

- Non-representativeness 

- Changing case definitions

- Testing inaccuracies (false negatives / positives)


#### Q3: Describe in your own words what the time-varying reprdocution number Rt represents? Further, list the data sources necessary to estimate the Rt? Focus on main manuscript and high level insights (ca. 20 minutes). Reading: Anne Cori, Neil M. Ferguson, Christophe Fraser, Simon Cauchemez, A New Framework and Software to Estimate Time-Varying Reproduction Numbers During Epidemics, American Journal of Epidemiology, Volume 178, Issue 9, 1 November 2013, Pages 1505–1512, https://doi.org/10.1093/aje/kwt133

<p>&nbsp;</p>

Model answer: 

- Date of onset of cases (if daily data are reported)

- Number of cases

- Serial interval distribution (the time between the onset of symptoms in a primary case and the onset of symptoms of secondary cases)


#### Q4: Downloading national COVID-19 cases data (England).
* Go to https://coronavirus.data.gov.uk/
* Download data option on the left
* Select Nation - England, latest (can’t find an option to select start and end dates) NewCasesBySpecimenDate, and copy the permanent link and use to load the data. `dat <- read.csv("InsertLinkHere")`

```{r, echo= FALSE}
dat <- read.csv("https://api.coronavirus.data.gov.uk/v2/data?areaType=nation&areaCode=E92000001&metric=newCasesBySpecimenDate&format=csv")
```

* For this practical we are only interested in cases between 1st October 2021 and 31st December 2021. So restrict your dates and sort them from October to Decemeber:

```{r}
dat <- dat %>%
  mutate(date = as.Date(date)) %>%
  arrange(date) %>%
  filter(date >= "2021-10-01", date <= "2021-12-31") %>%
  mutate(t_end = 1:n())
```


#### Q5: Visualising COVID-19 cases data.
```{r}
ggplot(data = dat, aes(x = date, y = newCasesBySpecimenDate)) +
  geom_point() +
  scale_x_date(date_breaks = "2 week") +
  theme_bw()

```
NOTE - In reality we would remove the weekend effect.

#### Q6: Estimating time-varying reproduction number, Rt.

We can run `estimate_R` on the incidence data to estimate the time-varying reproduction number R. For this, we need to specify i) the time window(s) over which to estimate Rt and ii) information on the distribution of the serial interval (SI).

For i), the default behavior is to estimate Rt over weekly sliding windows (for example: window 1 = 01Oct2021 to 07Oct2021, window 2 = 02Oct2021 to 08Oct2021, window 3 = 03Oct2021 to 09Oct2021, etc). This can be changed through the `t_start` and `t_end` arguments (see below, “Changing the time windows for estimation”). For ii), there are several options, specified in the method argument.

The simplest is the `parametric_si` method, where you only specify the mean and standard deviation of the SI.

In this example, we only specify the mean and standard deviation of the serial interval. In the following example, we use the mean (4 days) and standard deviation (1.5) of the serial interval for SARS-CoV-2 Delta variant of concern from XX Ferguson et al., Nature, 2005: https://www.nature.com/articles/nature04017 XX

Make sure the date are in chronological order, because `estimate_R` requires daily cases from earliest date.

After estimating Rt, plot the output. (ca. 10 minutes)

```{r Estimate Rt,echo=TRUE}
res_parametric_si <- estimate_R(dat$newCasesBySpecimenDate, method="parametric_si", config = make_config(list(mean_si = <xx>, std_si = <xx>)))

# Extract the estimates time-varying reproduction numbers
res_R <- res_parametric_si[["R"]]

# Add dates from the original dataset
res_R <- merge(res_R, dat, by = c("t_end"))

# Plot median and 95% uncertainty intervals
ggplot(data = res_R, aes(x = date, y = `Median(R)`)) +
  geom_line() +
  geom_ribbon(aes(ymin = `Quantile.0.025(R)`, ymax = `Quantile.0.975(R)`), alpha = 0.4) +
  geom_hline(yintercept = 1, color = "blue") +
  scale_x_date(date_breaks = "2 week") +
  theme_bw()
```


#### Q7: Based on the above plot, describe the epidemiological dynamics of the outbreak by referencing the change in the time-varying reproduction number, Rt. (ca. 10 minutes)

XX


#### Q8: Estimating variant-specific case incidence.

Between October and December 2021, the previously dominant Delta variant was rapidly displaced the Omicron variant. It is therefore important that we estimate the time-varying reproduction number for each variant separately. For this, we need to first estimate the case incidence attributed to each variant using lineage proportion estimates from genomic surveillance data. In this example, we will use publicably available from the Sanger Institute (https://covid19.sanger.ac.uk/lineages/raw). A processed version of this data has already been prepared for you and can be found in the data folder (daily_lineage_freqs.csv).

Visualise the data and describe what you see. (ca. 10 minutes)

```{r}
lineage_freqs.df <- read.delim('./daily_lineage_freqs.csv', sep=',') %>%
  mutate(date=as.Date(date))
ggplot() +
  geom_line(dat=lineage_freqs.df, aes(x=date, y=proportion, color=VOC)) +
  theme_bw()
```

To estimate the incidence of each variant separately, we have to multiple the daily case incidence (from Q1-2) by the proportion of cases attributed to each variant. Visualise the variant-specific incidence and describe what you see. (ca. 10 minutes)

```{r}
omicron.incidence.df <- merge(dat, lineage_freqs.df[lineage_freqs.df$VOC == 'Omicron',], by=c('date')) %>%
  mutate(estCases=newCasesBySpecimenDate*proportion)

delta.incidence.df <- merge(dat, lineage_freqs.df[lineage_freqs.df$VOC == 'Delta',], by=c('date')) %>%
  mutate(estCases=newCasesBySpecimenDate*proportion)

combined.incidence.df <- rbind(
  omicron.incidence.df %>% select(date, estCases) %>% mutate(VOC='Omicron'),
  delta.incidence.df %>% select(date, estCases) %>% mutate(VOC='Delta')
)

ggplot() +
  geom_point(dat=combined.incidence.df, aes(x=date, y=estCases, color=VOC)) +
  scale_x_date(date_breaks="2 week") +
  labs(x='date', y='estimated daily case incidence') +
  theme_bw()
```

Do you anticipate any potential issues if we are to estimate Rt for the same time period as in Q3?


#### Q9: Estimate variant-specific time-varying reproduction number, Rt.

Before we go ahead and run `estimate_R` on the variant-specific incidence data, we have to first make sure that the parameters we use to describe the distribution of the serial interval (SI) is appropriate for each variant. There are several studies that have estimated the SI for both the Delta and Omicron variants (e.g. Backer et al., Eurosurveillance, 2022; Águila-Mejía et al., Emerging Infectious Diseases, 2022). In this example, we will use a mean serial interval of 4.1 days (SD: 2.8 days) and 3.5 days (SD: 2.4 days) for the Delta and Omicron variants, respectively.

What is the significance of a shorter serial interval? (ca. 5 minutes)

Now having both the variant-specific incidence data and the SI parameters, we can estimate Rt for each variant separately by following the same steps as in Q3. Make sure to replace the first argument in `estimate_R` with the correct dataframe and column name. Visualise the results and describe what you see (change the limits of the x-axis as appropriate according your observations from Q5). (ca. 10 minutes)

```{r Estimate Rt,echo=TRUE}
omicron_res_parametric_si <- estimate_R(omicron.incidence.df$estCases, method="parametric_si", config = make_config(list(mean_si = xx, std_si = xx)))
# Extract the estimates time-varying reproduction numbers
omicron_res_R <- omicron_res_parametric_si[["R"]]
# Add dates from the original dataset
omicron_res_R <- merge(omicron_res_R, dat, by = c("t_end"))

delta_res_parametric_si <- estimate_R(delta.incidence.df$estCases, method="parametric_si", config = make_config(list(mean_si = xx, std_si = xx)))
# Extract the estimates time-varying reproduction numbers
delta_res_R <- delta_res_parametric_si[["R"]]
# Add dates from the original dataset
delta_res_R <- merge(delta_res_R, dat, by = c("t_end"))

# Combined dataframes
combined_res_R <- rbind(omicron_res_R %>% mutate(VOC='Omicron'), delta_res_R %>% mutate(VOC='Delta'))

# Plot median and 95% uncertainty intervals
ggplot(data = combined_res_R, aes(x = date, y = `Median(R)`, color = VOC)) +
  geom_line() +
  geom_ribbon(aes(ymin = `Quantile.0.025(R)`, ymax = `Quantile.0.975(R)`), alpha = 0.4) +
  geom_hline(yintercept = 1, color = "blue") +
  scale_x_date(date_breaks = "2 week", limits=c(as.Date('2021-10-01'), as.Date('2021-12-31'))) +
  scale_y_continuous(limits = c(0, 20)) +
  theme_bw()
```
 
Describe the patterns in the Rt estimates for each variant. Can you explain them? (ca. 10 minutes)


#### Q10: Accounting for case importation (Omicron variant).

We have so far assumed that all cases are linked by local transmission. However, it was later found through analysis of travel history data that, a substantial proportion of cases of the Omicron variant were in fact imported from abroad [see examples of analyses that make use of such data here: (1) xx, (2)]. A table containing the proportion of Omicron cases that were imported on each day is available and can be found in the data folder (daily_import_proportions.csv). We can use this information to account for case importation in our Rt estimates.

Use the data to estimate the number of Omicron cases that were local/imported on each day. Visualise the results and describe what you see. Can you explain the patterns? (ca. 15 minutes)

```{r}
import.proportion.df <- read.delim('./daily_import_proportions.csv', sep=',') %>%
  mutate(date=as.Date(date))

adjusted.incidence.df <- omicron.incidence.df %>%
  select(date, estCases) %>%
  mutate(importProportion = import.proportion.df$prop) %>%
  mutate(imported = as.integer(estCases*importProportion)) %>%
  mutate(local = estCases - importedCases)

ggplot() +
  geom_point(dat=adjusted.incidence.df, aes(x=date, y=local), color='black') +
  geom_point(dat=adjusted.incidence.df, aes(x=date, y=imported), color='red') +
  scale_x_date(date_breaks="2 week") +
  labs(x='date', y='estimated daily number of local/imported cases') +
  theme_bw()
```

Having adjusted for case importation, we can now re-estimate Rt for the Omicron variant following the same steps as before. Visualise both the original and adjusted Rt estimates and describe what you see (change the limits of the x-axis accordingly as before). Is there a difference following the adjustment, and if yes, can you explain why? (ca. 10 minutes)

```{r}
omicron_res_with_imports <- estimate_R(adjusted.incidence.df, method = "parametric_si", config = make_config(list(mean_si = <xx>, std_si = <xx>)))
omicron_res_with_imports_R <- omicron_res_with_imports[["R"]]
omicron_res_with_imports_R <- merge(omicron_res_with_imports_R, dat, by = c("t_end"))

ggplot() +
  geom_line(dat = omicron_res_R, aes(x = date, y = `Median(R)`), color='black') +
  geom_line(dat = omicron_res_with_imports_R, aes(x = date, y = `Median(R)`), color='red') +
  geom_ribbon(dat = omicron_res_R, aes(x = date, ymin = `Quantile.0.025(R)`, ymax = `Quantile.0.975(R)`), alpha = 0.4, fill='black', color='black') +
  geom_ribbon(dat = omicron_res_with_imports_R, aes(x = date, ymin = `Quantile.0.025(R)`, ymax = `Quantile.0.975(R)`), alpha = 0.4, fill='red', color='red') +
  geom_hline(yintercept = 1, color = "blue") +
  scale_x_date(date_breaks = "2 week", limits=c(as.Date('2021-10-01'), as.Date('2021-12-31'))) +
  scale_y_continuous(limits=c(0, 9)) +
  theme_bw()
```

### Q11: Name other sources of uncertainty in Rt estimation?

Model answer: 

- Under-reporting

- Spatial scale of reporting

- Health care seeking